### PR TITLE
8299079: Better interface nmethod oop accesses

### PR DESCRIPTION
**Pull request by fisk**
Today when oops are read from nmethods, we use the same interfacing that we would use to read any other IN_NATIVE oops. That isn't enough for generational ZGC. The nmethod oops are rather different from other IN_NATIVE oops in that they are encoded in the machine code as direct pointers. Normally, IN_NATIVE oops with generational ZGC are not direct pointers, but are colored pointered that can be converted to direct pointers by going through a load barrier. For nmethod oops, that conversion requires external knowledge from the surrounding nmethod. I propose a new NMethodAccess to interface better with the GC.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8299079](https://bugs.openjdk.org/browse/JDK-8299079): Better interface nmethod oop accesses

### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to [f9577ba8](https://git.openjdk.org/jdk/pull/11738/files/f9577ba89b8a7bc37ea8d7e5193a3b6d2f035ade)
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to [f9577ba8](https://git.openjdk.org/jdk/pull/11738/files/f9577ba89b8a7bc37ea8d7e5193a3b6d2f035ade)

### Contributors
 * Axel Boldt-Christmas `<aboldtch@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11738/head:pull/11738` \
`$ git checkout pull/11738`

Update a local copy of the PR: \
`$ git checkout pull/11738` \
`$ git pull https://git.openjdk.org/jdk pull/11738/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11738`

View PR using the GUI difftool: \
`$ git pr show -t 11738`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11738.diff">https://git.openjdk.org/jdk/pull/11738.diff</a>

</details>

### CODE REVIEW

**Review by vnkozlov — APPROVED** [review]
Good.

File: src/hotspot/share/code/nmethod.cpp [hunk 1]
```diff
@@ -1462,14 +1462,16 @@ oop nmethod::oop_at(int index) const {
   if (index == 0) {
     return NULL;
   }
-  return NativeAccess<AS_NO_KEEPALIVE>::oop_load(oop_addr_at(index));
+
```
Comment (dholmes-ora) [inline src/hotspot/share/code/nmethod.cpp:1465]: Nit: the extra blank lines are unnecessary.

File: src/hotspot/share/oops/access.hpp [hunk 2]
```diff
@@ -285,6 +285,11 @@ class HeapAccess: public Access<IN_HEAP | decorators> {};
 template <DecoratorSet decorators = DECORATORS_NONE>
 class NativeAccess: public Access<IN_NATIVE | decorators> {};
 
+// Helper for performing accesses in nmethods. These accesses
+// may resolve an accessor on a GC barrier set
```
Comment (dholmes-ora) [inline src/hotspot/share/oops/access.hpp:289]: Nit: period at end of sentence please.

**Review by dholmes-ora — APPROVED** [review]
A couple of minor nits, but the new API looks fine. Thanks.

**Comment by fisk** [comment]
Thanks for the reviews, @vnkozlov and @dholmes-ora! Will fix the nits before integrating.

**Pull request closed without merging**